File: src/services/tools/web/newsSearch.ts
import { z } from 'zod';
import type { ToolDef } from '../types';
import { WebSearchService } from '../../WebSearchService';

export const newsSearchTool: ToolDef = {
  name: 'news_search',
  description: 'Search for recent news articles. Use when the user asks about current events, recent developments, or time-sensitive topics.',
  category: 'web',
  inputSchema: z.object({
    query: z.string().describe('The news search query.'),
    maxResults: z.number().min(1).max(5).optional().default(5).describe('Number of news results (1–5).'),
    freshness: z.enum(['hour', 'day', 'week', 'month']).optional().default('week').describe('How recent the news should be.'),
  }),
  execute: async ({ query, maxResults, freshness }) => {
    try {
      const result = await WebSearchService.newsSearch({ query, maxResults: Math.min(maxResults ?? 5, 5), freshness });
      return {
        results: result.results.map(r => ({
          title: r.title, url: r.url, snippet: r.snippet,
          publishedDate: r.publishedDate, source: r.source,
        })),
        totalResults: result.totalResults,
      };
    } catch (error: any) {
      throw new Error(error.message || 'News search failed.');
    }
  },
};

File: agent/internal/tool/news_search.go
package tool

import (
	"context"
	"fmt"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func newsSearchTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "news_search",
			Description: "Search for recent news articles on a topic. Returns headlines, sources, and publication dates.",
			Category:    "research",
			Parameters: map[string]api.ParamDef{
				"query":       {Type: "string", Description: "The news search query", Required: true},
				"max_results": {Type: "number", Description: "Max articles (1-10)", Required: false, Default: 5},
				"freshness":   {Type: "string", Description: "Time range for news", Required: false, Default: "week", Enum: []string{"hour", "day", "week", "month"}},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			query, _ := params["query"].(string)
			if query == "" {
				return nil, fmtError("query is required")
			}
			maxResults, _ := params["max_results"].(float64)
			if maxResults == 0 {
				maxResults = 5
			}
			freshness, _ := params["freshness"].(string)
			if freshness == "" {
				freshness = "week"
			}

			return postToExpress(ctx, e.expressURL, "newsSearch", map[string]any{
				"query":      query,
				"maxResults": int(maxResults),
				"freshness":  freshness,
			})
		},
	}
}
